mne/preprocessing/_csd.py

```diff
@@ -0,0 +1,168 @@
+    sgi = np.sum(TC)  # compute sum total
+
+    return Gi, TC, sgi
+
+
+def _compute_csd(data, G_precomputed, H, radius):
+    """Compute the CSD."""
+    n_channels, n_times = data.shape
+    mu = data.mean(0)[None]
+    Z = data - mu
+    X = np.zeros_like(data)
+    radius **= 2
+
+    Gi, TC, sgi = G_precomputed
+
+    Cp2 = np.dot(Gi, Z)
+    c02 = np.sum(Cp2, axis=0) / sgi
+    C2 = Cp2 - np.dot(TC[:, None], c02[None, :])
+    X = np.dot(C2.T, H).T / radius
+    return X
+
+
+def compute_current_source_density(inst, lambda2=1e-5, stiffness=4,
+                                   n_legendre_terms=50,
+                                   sphere=(0., 0., 0., 0.095),
+                                   copy=True):
+    """Get the current source density (CSD) transformation.
+
+    Transformation based on spherical spline surface Laplacian.
+
+    .. note:: This function applies an average reference to the data.
+              Do not transform CSD data to source space.
+
+    Parameters
+    ----------
+    inst : instance of Raw, Epochs or Evoked
+        The data to be transformed.
+    lambda2 : float
+        Regularization parameter, produces smoothnes. Defaults to 1e-5.
+    stiffness : float
+        Stiffness of the spline. Also referred to as `m`
+        (if g_matrix or h_matrix is provided this will be ignored).
+    n_legendre_terms : int
+        Number of Legendre terms to evaluate (if g_matrix or h_matrix
+        is provided this will be ignored).
+    sphere : tuple, (float, float, float, float)
+        The sphere, head-model of the form (x, y, z, r) where x, y, z
+        is the center of the sphere and r is the radius.
+    copy : bool
+        Whether to overwrite instance data or create a copy.
+
+    Returns
+    -------
+    inst_csd : instance of Epochs or Evoked
+        The transformed data. Output type will match input type.
+    """
+    _validate_type(inst, (BaseEpochs, BaseRaw, Evoked), 'inst')
+
+    if 'csd' in inst.info['comps']:
+        raise ValueError('CSD already applied, should not be reapplied')
+
+    inst = inst.copy() if copy else inst
+
+    picks = pick_types(inst.info, meg=False, eeg=True, exclude='bads')
+
+    if len(picks) == 0:
+        raise ValueError('No EEG channels found.')
+
+    if lambda2 is None:
+        lambda2 = 1e-5
+
+    _validate_type(lambda2, (float, int), 'lambda2')
+    if 0 > lambda2 or lambda2 > 1:
+        raise ValueError('lambda2 must be between 0 and 1, got %s' % lambda2)
+
+    _validate_type(stiffness, (float, int), 'stiffness')
+    if stiffness < 0:
+        raise ValueError('stiffness must be non-negative got %s' % stiffness)
+
+    _validate_type(n_legendre_terms, (int), 'n_legendre_terms')
+    if n_legendre_terms < 1:
+        raise ValueError('n_legendre_terms must be greater than 0, '
+                         'got %s' % n_legendre_terms)
+
+    _validate_type(sphere, tuple, 'sphere')
+    x, y, z, radius = sphere
+    _validate_type(x, float, 'x')
```

Comment:
these can just be `'numeric'`, otherwise someone will pass `(0, 0, 0, 0.08)` and get an unexpected error here because `0` is an `int` not a `float`